include/emp/tools/string_utils.hpp

```diff
@@ -129,6 +131,58 @@ namespace emp {
     return web_safe;
   }
 
+
+  /// Returns url encoding of value.
+  /// See https://en.wikipedia.org/wiki/Percent-encoding
+  // adapted from https://stackoverflow.com/a/17708801
+  template<bool encode_space=false>
+  std::string url_encode(const std::string &value) {
+    std::ostringstream escaped;
+    escaped.fill('0');
+    escaped << std::hex;
+
+    for (const auto c : value) {
+
+      // If encoding space, replace with +
+      if ( encode_space && c == ' ' ) escaped << '+';
+      // Keep alphanumeric and other accepted characters intact
+      else if (
+        std::isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~'
+      ) escaped << c;
+      // Any other characters are percent-encoded
+      else {
+        escaped << std::uppercase;
+        escaped << '%' << std::setw(2) << (static_cast<int>(c) & 0x000000FF);
+        escaped << std::nouppercase;
+      }
+
+    }
+
+    return escaped.str();
+  }
+
+  /// Returns url decoding of string.
+  /// See https://en.wikipedia.org/wiki/Percent-encoding
+  // adapted from https://stackoverflow.com/a/29962178
+  template<bool decode_plus=false>
+  std::string url_decode(const std::string& str){
+    std::string res;
+
+    for (size_t i{}; i < str.size(); ++i) {
```

Comment:
Range based for loop?

Reply:
Since we might encounter percent encoded triplets `%XX` it's easiest to keep it like this.